common/src/main/java/org/jboss/pnc/common/gerrit/ScmUrlGeneratorProvider.java

```diff
@@ -0,0 +1,131 @@
+
+    /**
+     * Returns an SCM url generator specific to a type of SCM
+     *
+     * @param provider type of SCM provider (Gerrit, Gitlab)
+     * @return generator of urls specific to the type of provider
+     * @throws ScmException throws if provider is unknown
+     */
+    public ScmUrlGenerator getScmUrlGenerator(@NotNull SCMProvider provider) throws ScmException {
+        var generator = scmUrlProviders.get(provider);
+        if (generator == null) {
+            throw new ScmException("Unknown SCM provider: " + provider);
+        }
+        return generator;
+    }
+
+    /**
+     * Simple heuristic logic for trying to determine what is the SCM provider out of url used to clone the build and
+     * internal representation of scm repo from the SCMRepository entity.
+     *
+     * @param scmUrl url used to clone repo in a Build (URI only)
+     * @param internalScmUrl internalUrl from SCMRepository of the Build (can be URI or SCP-like)
+     * @return SCMProvider that is likely to have been used
+     * @throws ScmException if the arguments are not appropriate URLs
+     */
+    public SCMProvider determineScmProvider(String scmUrl, String internalScmUrl) throws ScmException {
+        URI scmUri;
+        try {
+            scmUri = new URI(scmUrl);
+        } catch (URISyntaxException e) {
+            throw new ScmException("Cannot parse url of " + scmUrl);
+        }
+
+        // simple logic based on host name
+        if (scmUri.getHost().contains("gitlab")) {
+            return SCMProvider.GITLAB;
+        } else if (scmUri.getHost().contains("gerrit") || scmUri.getPath().startsWith("/gerrit/")) {
+            return SCMProvider.GERRIT;
+        }
+
+        // try some logic from internalScmUrl
+        if (isScpLike(internalScmUrl)) {
+            return determineFromScpUri(internalScmUrl);
+        } else {
+            return determineFromUri(internalScmUrl);
+        }
+    }
+
+    private static SCMProvider determineFromScpUri(String internalScmUrl) throws ScmException {
+        try {
+            GitSCPUrl scpUrl = GitSCPUrl.parse(internalScmUrl);
+
+            if (scpUrl.getHost().contains("gitlab")) {
+                return SCMProvider.GITLAB;
+            } else if (scpUrl.getHost().contains("gerrit")) {
+                return SCMProvider.GERRIT;
+            }
+
+            // gitlab urls tend to be SCP like
+            return SCMProvider.GITLAB;
+        } catch (MalformedURLException e) {
+            throw new ScmException("Cannot parse url of " + internalScmUrl);
+        }
+    }
+
+    private static SCMProvider determineFromUri(String internalScmUrl) throws ScmException {
+        try {
+            URI internalUri = new URI(internalScmUrl);
+
+            if (internalUri.getHost().contains("gitlab")) {
+                return SCMProvider.GITLAB;
+            } else if (internalUri.getHost().contains("gerrit")) {
+                return SCMProvider.GERRIT;
+            }
```

Comment:
Isn't this just the copy of above?

Reply:
`internalScmUrl` have no `/gerrit/` in the path because they tend to be in form of `git+ssh://` so no. If you mean from `determineFromScpUri`, then it's still a different because I am dealing with different classes (`URI` vs `GitSCPUrl`).

Reply:
Right, I didn't realize you are now using two different urls to get the type. Why is that? Shouldn't the scmUrl be enough?

Reply:
For gitlab urls, I don't want to rely on just the fact that there is gitlab in the host (like we do not have for gerrit). We have a second way how to discern gitlab which would be that it strictly favours scp-like urls. The format is never present `scmUrl` so that's why I also look at SCMRepository.